[MOO-494] Add docs for dangling references (#4068)
mendix/docs: content/refguide/offline-first.md

### 2.2 Synchronization Phases

> The synchronization process consists of two phases. In the [upload phase](#upload), your app updates the server database with the new or changed objects that are committed. In the [download phase](#download), your app updates its local database using data from the server database. Note that synchronization only works on the database level.

Author: ✔️

#### 2.2.1 Upload Phase {#upload}

The upload phase executes the following operations:

> 1. As the local database can be modified only by committing or deleting an object, such an object can be either a new object created while offline, or an existing object previously synced from the server. The upload phase detects which objects have been committed to the local database since the last synchronization. The detection logic differs per synchronization type. For Synchronize all, all committed objects in the local database are checked. For Synchronize objects, all committed objects from the list of selected objects are checked.

Reviewer: > all committed objects from the list of selected objects are checked.

This is a leftover from dangling reference description: we don't "check" anything anymore. But we can say "included" to mention which objects are selected for synchronization. 

We also don't seem to mention "Synchronize unsynced objects" mode.

Author: It's not about dangling in any way, it's about selecting what data to synchronize indeed. If you find "checked" controversial, we can replace it with "selected for synchronization" for instance.

Author: "Check" here means "scan the db and pick what needs to be synced". "Checked" and "included" for sync are not the same sets: an object can be "checked" (entire database in case of full sync) but not "included" (as we include only dirty) for sync.

Alternalively, we can say something like "for full sync all ditry objects are included". What do you think?

Reviewer: Indeed I misunderstood the usage of the check here - with your explanation is makes sense. I'd keep it as-is.

2. There might be objects deleted from the device database since the last synchronization. The upload phase checks which objects have been deleted.

> 3. If there are any changed or new file objects, their content is uploaded to the server and stored there temporarily. Each file is uploaded in a separate network request.

Reviewer 2: Same as below; I am not sure why the separate network request is usefull if left unexplained in this context.

Reviewer: We can continue with: "If upload of a file fails, the whole synchronization is aborted without causing any changes in both the server and local database.

Author: ✔️

> 4. All the changed and new objects are sent to the server, and the content of the files is linked to the objects. The server performs referential integrity validation of the objects (see the [Dangling references](#dangling-references) section for more info). The objects are committed to the server database. Information about deleted objects is also sent to the server so the server can delete them from its database too. This step is performed in a single network request.

Reviewer 2: I miss the significance of why a single network request is used. I am not sure we want to go into detail, but without extra info it's not helpful to a reader?

Author: A single network request implies that the operation is kind of atomic.
There's also a difference between upload and download, where for the latter we use one request per entity.

I do feel we can elaborate more on the differences between single/multiple requests, but I don't really know where to draw the line to be honest.

Reviewer 2: Exactly that is my point. We know that; will a customer understand that?

Reviewer 2: I think a more explanatory sentence; like the operation happens in one request and that's the reason why, similar to Ekrem's suggestion is aggreable with me.

> 5. Any configured before- or after-commit or before- or after-delete event handlers on these objects will run on the server as usual: after the data has been uploaded and before it is downloaded.

Reviewer: * We also execute create event handlers for new objects.
* "... and before local database is updated. This means any further changes you make to the synchronised objects in the event handlers will be updated in the local database in the download phase"

There's one exception to the description above - changing the contents of a file entities aren't reflected back. Maybe important to mention.

Author: Nice catch Ekrem 👍 Added.

* The object is not valid according to domain-level validation rules
* The object has a dangling reference (see the [Dangling references](#dangling-references) section for more info)

{{% alert type="warning" %}}When a synchronization error occurs because of one the reasons above, an object's commit is skipped, its changes are ignored, and references from other objects to it become invalid. Objects referencing such a skipped object (which are not triggering errors) will be synchronized normally. Such a situation is likely to be a modeling error and is logged on the server. To prevent data loss, the attribute values for such objects are stored in the `System.SynchronizationError` entity (since Mendix 8.12).  {{% /alert %}}

#### 2.6.3 Dangling references {#dangling-references}

During the synchronization the server performs referential integrity validation of the new or changed objects that are being synchronized to the server.
This validation ensures that none of the synchronized objects have associations pointing to an object that exists only in the device database.
If an association doesn’t satisfy this condition, this is called a dangling reference.

> For example, when a committed `City` object refers to an uncommitted `Country` object, synchronizing the `City` object alone will yield an invalid `Country` object reference, which is a dangling reference error.

Reviewer 2: Maybe simplify the language more. Does commit has the meaning we thing for a user?
In offline context, commit means saving to the database, do we actually verify danglin references then? 

Is commit the right verb to describe the operation of sync and saving to the database?

Author: > do we actually verify danglin references then?

The validation happens upon sync only. The wording here is a bit misleading, good point.

> Is commit the right verb to describe the operation of sync and saving to the database?

We call it "commit" everywhere in the docs, there's a nanoflow activity called "Commit", etc. 
That's the wording we use. The concept of committing an object is known to those who read the docs. 
We also explicitly mention that sync works at the database level.

Reviewer: "Commit" is the correct word to use indeed: intended as "saving to the offline db" and here specifies the state of the objects. Dangling reference check does not happen at commit time, but when the committed object is synchronised.

> A dangling reference error is a modeling error.

Reviewer 2: These deserve their own sections. Could we split dangling references and how to avoid them to their own page?

Reviewer 2: Maybe a database integrity section or something

Reviewer: +1 for a separate section, but we have more stuff to put under such a section. That's why I'd do it in a separate story.

> To prevent them from happening during selective synchronization make sure you either select both sides of the association for the synchronization

Reviewer: The suggestion of selecting both objects wouldn't work because `Country` is stated as uncommitted above.

Author: Good point. Fixed ✔️

or make sure the `Country` object is synchronized before synchronizing `City`.
To prevent dangling reference errors during full synchronization make sure both sides of the association are committed before synchronizing.

When some of the synchronized objects have dangling references, 
the server will synchronize all other objects except the ones with dangling references,
for each of which it will create a synchronization error and store it in
`System.SynchronizationError` entity. The error message in such cases looks like the following:

**Synchronizing an object of type City with guid ... has failed due to a modelling error. 
The object has a reference to other objects (City_Country) that have not been synchronised to the runtime yet. 

> This breaks referential integrity of the object because it references a non-existing object in the runtime database. 

Reviewer: ```suggestion
This breaks referential integrity of the object because it references an object that only exists on the device database.
```

Author: This text is exactly what the runtime logs in the console. Do you suggest to change it there as well? The existing version is fine with me, as it's the message shown by the runtime, so "non-existing" is not ambiguous in this case.

Reviewer: Yes, I'd also change it on the logs as well - but noticed that my version needs adjustment - we need the drop "database" from it because a dangling object can be a new one that has not yet been committed to the offline database.

Author: Let's do this as a separate change set after we change the message in the runtime then. Created a small story.

* A member of the object has become inaccessible due to access rules
* An error occurs during the execution of a before- or after-commit event microflow
* The object is not valid according to domain-level validation rules
* The object has a dangling reference (see the [Dangling references](#dangling-references) section for more info)

{{% alert type="warning" %}}When a synchronization error occurs because of one the reasons above, an object's commit is skipped, its changes are ignored, and references from other objects to it become invalid. Objects referencing such a skipped object (which are not triggering errors) will be synchronized normally. Such a situation is likely to be a modeling error and is logged on the server. To prevent data loss, the attribute values for such objects are stored in the `System.SynchronizationError` entity (since Mendix 8.12).  {{% /alert %}}

#### 2.6.3 Dangling references {#dangling-references}

During the synchronization the server performs referential integrity validation of the new or changed objects that are being synchronized to the server.

> This validation ensures that none of the synchronized objects have associations pointing to an object that exists only in the device database.

Reviewer: I think we should drop "database" from the end of the sentence.

Author: ✔️
